File: Java/008_String to Integer (atoi).java
class Solution {
    public int myAtoi(String str) {
        int index = 0;
        while (index < str.length() && Character.isSpaceChar(str.charAt(index)))
            index++;
        boolean sign = true;
        if (index < str.length()) {
            switch (str.charAt(index)) {
                case '-' : sign = false;
                case '+' : index++;
                default : break;
            }
        }
        long res = 0;
        while (index < str.length() && Character.isDigit(str.charAt(index))) {
            res = res * 10 + str.charAt(index) - '0';
            if (sign && res > Integer.MAX_VALUE)
                return Integer.MAX_VALUE;
            if (!sign && -res < Integer.MIN_VALUE)
                return Integer.MIN_VALUE;
            index++;
        }
        return sign ? (int)res : (int)-res;
    }
}

public class Test {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.myAtoi("   +2147483647"));
        System.out.println(s.myAtoi("-222222222222222"));
        System.out.println(s.myAtoi("-100"));
    }
}

File: Python/008_String to Integer (atoi).py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        str = str.strip()
        l, res = len(str), 0
        if l == 0:
            return 0
        flag = True if str[0] == '-' else False
        i = 1 if str[0] == '-' or str[0] == '+' else 0
        od0, od9 = ord('0'), ord('9')
        while i < l:
            t = ord(str[i])
            if t < od0 or t > od9:
                break
            res = res * 10 + t - od0
            if res > 2147483648:
                break
            i += 1
        if flag and -res < -2147483648:
            res = 2147483648
        if not flag and res > 2147483647:
            res = 2147483647
        return -res if flag else res